clustered-indexes.md

```diff
@@ -0,0 +1,188 @@
+
+## Default behavior
+
+TiDB has always supported clustered indexes, provided the following criteria are true:
+
+- The table contains a `PRIMARY KEY`
+- The primary key is an `INTEGER` or `BIGINT`
+- The primary key consists of only one column
+
+When any of these criteria are not met, TiDB will create a hidden 64-bit `handle` to organize the table. Querying table rows by a clustered index is more efficient than by a non-clustered index because it can be completed as a single step. The following `EXPLAIN` output compares a table that supports clustered indexes with one that does not:
+
+```sql
+CREATE TABLE always_clusters_in_all_versions (
+ id BIGINT NOT NULL PRIMARY KEY auto_increment,
+ b CHAR(100),
+ INDEX(b)
+);
+
+CREATE TABLE does_not_cluster_by_default (
+ guid CHAR(32) NOT NULL PRIMARY KEY,
+ b CHAR(100),
+ INDEX(b)
+);
+
+INSERT INTO always_clusters_in_all_versions VALUES (1, 'aaa'), (2, 'bbb');
+INSERT INTO does_not_cluster_by_default VALUES ('02dd050a978756da0aff6b1d1d7c8aef', 'aaa'), ('35bfbc09cb3c93d8ef032642521ac042', 'bbb');
+
+EXPLAIN SELECT * FROM always_clusters_in_all_versions WHERE id = 1;
+EXPLAIN SELECT * FROM does_not_cluster_by_default WHERE guid = '02dd050a978756da0aff6b1d1d7c8aef';
+```
+
+```sql
+Query OK, 0 rows affected (0.09 sec)
+
+Query OK, 0 rows affected (0.10 sec)
+
+Records: 2  Duplicates: 0  Warnings: 0
+
+Records: 2  Duplicates: 0  Warnings: 0
+
++-------------+---------+------+---------------------------------------+---------------+
+| id          | estRows | task | access object                         | operator info |
++-------------+---------+------+---------------------------------------+---------------+
+| Point_Get_1 | 1.00    | root | table:always_clusters_in_all_versions | handle:1      |
++-------------+---------+------+---------------------------------------+---------------+
+1 row in set (0.00 sec)
+
++-------------+---------+------+--------------------------------------------------------+---------------+
+| id          | estRows | task | access object                                          | operator info |
++-------------+---------+------+--------------------------------------------------------+---------------+
+| Point_Get_1 | 1.00    | root | table:does_not_cluster_by_default, index:PRIMARY(guid) |               |
++-------------+---------+------+--------------------------------------------------------+---------------+
+1 row in set (0.00 sec)
+```
+
+The `EXPLAIN` results look similar, but in the second example TiDB must first read the `PRIMARY KEY` index on `guid` in order to find the `handle` value. This is more obvious in the following example where the value of the `PRIMARY KEY` is not in the index on `does_not_cluster_by_default.b`. An extra lookup must be performed on the table rows (`└─TableFullScan_5`) to convert the `handle` to the `PRIMARY KEY` value of `guid`:
```

Comment:
```suggestion
The two `EXPLAIN` results above look similar, but in the second example, TiDB must first read the `PRIMARY KEY` index on the `guid` column in order to find the `handle` value. This is more obvious in the following example where the `PRIMARY KEY` value is not in the index on `does_not_cluster_by_default.b`. TiDB must perform an extra full table scan on the table rows (`└─TableFullScan_5`) to convert the `handle` value to the `PRIMARY KEY` value of `guid`:
```

Reply:
I don't have permissions to partially accept your edit, but it's not correct. It is an additional lookup, not an extra full table scan.

Reply:
Comment changed. PTAL again, thanks~